tarantool/dbapi.py

```diff
@@ -0,0 +1,251 @@
+# -*- coding: utf-8 -*-
+from tarantool.connection import Connection as BaseConnection
+from tarantool.error import *
+
+
+paramstyle = 'named'
+apilevel = "2.0"
+threadsafety = 1
+
+
+class Cursor:
+
+    def __init__(self, conn):
+        self._c = conn
+        self._lastrowid = None
+        self._rowcount = None
+        self.arraysize = 1
+        self._rows = None
+
+    def callproc(self, procname, *params):
+        """
+        Call a stored database procedure with the given name. The sequence of
+        parameters must contain one entry for each argument that the
+        procedure expects. The result of the call is returned as modified
+        copy of the input sequence. Input parameters are left untouched,
+        output and input/output parameters replaced with possibly new values.
+        """
+        raise NotSupportedError("callproc() method is not supported")
+
+    @property
+    def rows(self):
+        return self._rows
+
+    @property
+    def description(self):
+        # FIXME Implement this method please
+        raise NotImplementedError("description() property is not implemented")
+
+    def close(self):
+        """
+        Close the cursor now (rather than whenever __del__ is called).
+        The cursor will be unusable from this point forward; DatabaseError
+        exception will be raised if any operation is attempted with
+        the cursor.
+        """
+        self._c = None
+        self._rows = None
+        self._lastrowid = None
+        self._rowcount = None
+
+    def _check_not_closed(self, error=None):
+        if self._c is None:
+            raise InterfaceError(error or "Can not operate on a closed cursor")
+        if self._c.is_closed():
+            raise InterfaceError("The cursor can not be used "
+                                 "with a closed connection")
+
+    def execute(self, query, params=None):
+        """
+        Prepare and execute a database operation (query or command).
+        """
+        self._check_not_closed("Can not execute on closed cursor.")
+
+        response = self._c.execute(query, params)
+
+        self._rows = response.data
+        self._rowcount = response.affected_row_count or -1
+        if response.autoincrement_ids:
+            self._lastrowid = response.autoincrement_ids[-1]
+        else:
+            self._lastrowid = -1
```

Comment:
Should not we set None, -1, -1 in case of an error? The standard says:

> The attribute is -1 in case no .execute*() has been performed on the cursor or the rowcount of the last operation is cannot be determined by the interface.

Technically a failed operation is still an operation, so may we should. What other implementations do?

Reply:
I agree that using `None` fits better to this case because it is set to `None` by default. Other connectors behave the same.

Reply:
I looked at the [standard](https://www.python.org/dev/peps/pep-0249/#lastrowid) and found that your change is right.

However my initial comment was about the different case. What if `self._c.execute(query, params)` will fail and raise an exception? Please, look at my comment in this context.

Reply:
I looked at [PyMysql implementation](https://github.com/PyMySQL/PyMySQL/blob/37fd1e1b0126d75d80eef59c053f80634b09bd75/pymysql/cursors.py#L322) and it seems they do the same as we do. If the query execute fails, the values of rowcount and lastrowid remain initial.

Reply:
Okay.